core/arch/arm/kernel/secure_partition.c

```diff
@@ -375,6 +386,146 @@ static TEE_Result sp_open_session(struct sp_session **sess,
 	return TEE_SUCCESS;
 }
 
+static TEE_Result sp_dt_get_u64(const void *fdt, int node, const char *property,
+				uint64_t *value)
+{
+	int len = 0;
+	const fdt64_t *cuint64 = NULL;
+
+	cuint64 = fdt_getprop(fdt, node, property, &len);
+	if (!cuint64 || len != sizeof(*cuint64))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	*value = fdt64_to_cpu(*cuint64);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result sp_dt_get_u32(const void *fdt, int node, const char *property,
+				uint32_t *value)
+{
+	int len = 0;
+	const fdt32_t *cuint32 = NULL;
+
+	cuint32 = fdt_getprop(fdt, node, property, &len);
+	if (!cuint32 || len != sizeof(*cuint32))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	*value = fdt32_to_cpu(*cuint32);
+	return TEE_SUCCESS;
+}
+
+static TEE_Result handle_fdt_dev_regions(struct sp_ctx *ctx, void *fdt)
+{
+	int node = 0;
+	int subnode = 0;
+	TEE_Result res = TEE_SUCCESS;
+	const char *dt_device_match_table = {
+		"arm,ffa-manifest-device-regions",
+	};
+
+	/*
+	 * Device regions are optional in the SP manifest, it's not an error if
+	 * we don't find any
+	 */
+	node = fdt_node_offset_by_compatible(fdt, 0, dt_device_match_table);
+	if (node < 0)
+		return TEE_SUCCESS;
+
+	fdt_for_each_subnode(subnode, fdt, node) {
+		uint64_t base_addr = 0;
+		uint32_t pages_cnt = 0;
+		uint32_t attributes = 0;
+		struct mobj *m = NULL;
+		bool is_nsec = false;
+		uint32_t perm = 0;
+		vaddr_t va = 0;
+		unsigned int idx = 0;
+
+		/*
+		 * Physical base address of a device MMIO region.
+		 * Currently only physically contiguous region is supported.
+		 */
+		if (sp_dt_get_u64(fdt, subnode, "base-address", &base_addr)) {
+			EMSG("Mandatory field is missing: base-address");
+			return TEE_ERROR_BAD_FORMAT;
+		}
+
+		/* Total size of MMIO region as count of 4K pages */
+		if (sp_dt_get_u32(fdt, subnode, "pages-count", &pages_cnt)) {
```

Comment:
"pages count" sounds a bit strange, are you sure it's not supposed to be "page count"?

Reply:
It's a bit strange indeed, this is how it's defined in the DT binding document.

Reply:
OK